### PR TITLE
Stream Mirrors, Sources and Updates Support

### PR DESCRIPTION
**Pull request by derekcollison**
Also various fixes, stability improvements, etc.

Signed-off-by: Derek Collison <derek@nats.io>

/cc @nats-io/core

### CODE REVIEW

File: server/const.go [hunk 1]
```diff
@@ -1,202 +0,0 @@
-// Copyright 2012-2021 The NATS Authors
```
Comment (ripienaar) [inline server/const.go:1]: this is probably a mistake right?

File: server/jetstream_api.go [hunk 2]
```diff
@@ -1020,6 +1024,55 @@ func (s *Server) jsStreamCreateRequest(sub *subscription, c *client, subject, re
...
+			s.sendAPIErrResponse(ci, acc, subject, reply, string(msg), s.jsonResponse(&resp))
+			return
+		}
+		if cfg.Mirror.FilterSubject != _EMPTY_ {
+			resp.Error = &ApiError{Code: 400, Description: "stream mirrors can not contain filtered subjects"}
+			s.sendAPIErrResponse(ci, acc, subject, reply, string(msg), s.jsonResponse(&resp))
+			return
+		}
+		if cfg.Mirror.OptStartSeq > 0 && cfg.Mirror.OptStartTime != nil {
+			resp.Error = &ApiError{Code: 400, Description: "stream mirrors can not have both start seq and start time configured"}
```
Comment (ripienaar) [inline server/jetstream_api.go:1045]: hmm, so then stream mirrors can in fact have no config at all, so the *StreamSource only contributes Name at that point, are we keeping *StreamSource there as a nod to the future when we somehow point to another account in there too?

File: server/const.go (hunk 1 above)
Comment (derekcollison) [reply server/const.go:1]: yes moved.

File: server/jetstream_api.go (hunk 2 above)
Comment (derekcollison) [reply server/jetstream_api.go:1045]: Not following..

**Pull request closed without merging**